### PR TITLE
[DEX-334] feat: algolia objects update

### PR DESCRIPTION
**Pull request by loicsay**
## Description:
The objects update command is available, allowing the users to partially update the objects in their indices.

## Notes 
- Can't make [`autoGenerateObjectIDIfNotExist`](https://www.algolia.com/doc/api-reference/api-methods/partial-update-objects/#method-param-autogenerateobjectidifnotexist) work. Got this error: "could not generate intermediate batch: missing `objectID` field in object"

### CODE REVIEW

File: pkg/cmd/objects/update/update.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	if err != nil {
+		return err
+	}
+
+	index := client.InitIndex(opts.Index)
+
+	options := []interface{}{
+		opt.CreateIfNotExists(opts.CreateIfNotExists), opt.AutoGenerateObjectIDIfNotExist(opts.AutoGenerateObjectIDIfNotExist),
+	}
+	_, err = index.PartialUpdateObjects(objectsToUpdate, options)
```
Comment (loicsay) [inline pkg/cmd/objects/update/update.go:123]: Can't make [autoGenerateObjectIDIfNotExist](https://www.algolia.com/doc/api-reference/api-methods/partial-update-objects/#method-param-autogenerateobjectidifnotexist) work. Got this error: "could not generate intermediate batch: missing objectID field in object"

@clemfromspace Do I do something wrong here? I tested the code from the client but couldn't make it work as expected 😞

File: pkg/cmd/objects/update/object.go [hunk 2]
```diff
@@ -1,78 +1,78 @@
...
-				objectToUpdate["objectID"] = valueString
+	// Each field could be either an `search.PartialUpdateOperation` or a scalar value
+	for k, v := range *o {
+		switch v := v.(type) {
+		case map[string]interface{}:
+			var op search.PartialUpdateOperation
+			if err := mapstructure.Decode(v, &op); err != nil {
+				return err
 			}
+			// Check thate the operation is valid
```
Comment (loicsay) [inline pkg/cmd/objects/update/object.go:69]:
```suggestion
// Check that the operation is valid
```

File: pkg/cmd/objects/update/object.go [hunk 3]
```diff
@@ -1,78 +1,78 @@
...
 	"encoding/json"
 	"fmt"
 
 	"github.com/algolia/algoliasearch-client-go/v3/algolia/search"
 	"github.com/algolia/cli/pkg/utils"
+	"github.com/mitchellh/mapstructure"
 )
 
-type ObjectsToUpdate []map[string]interface{}
+// Object is a map[string]interface{} that can be unmarshalled from a JSON object
```
Comment (loicsay) [inline pkg/cmd/objects/update/object.go:12]: Good habit to add comments to describe `Object` type for such a generic word, I'll try to do the same in the future 🔥

File: pkg/cmd/objects/update/object.go [hunk 4]
```diff
@@ -1,78 +1,78 @@
...
-					mapValue, isMap := rawValue.(map[string]interface{})
-					if !isMap {
-						return fmt.Errorf("Object value/operation not recognized for object %d", objectIndex+1)
-					}
-					mapValueString, isMapValueString := mapValue["value"].(string)
-					// JSON unmarshal numbers to float64 by default
-					mapValueFloat, isMapValueFloat := mapValue["value"].(float64)
-					mapValueInt := int(mapValueFloat)
+	// The object must be a map[string]interface{}
+	switch v := v.(type) {
```
Comment (loicsay) [inline pkg/cmd/objects/update/object.go:49]: TIL 💡

File: pkg/cmd/objects/update/object.go [hunk 5]
```diff
@@ -1,78 +1,78 @@
...
+	Increment     string = "Increment"
+	Decrement     string = "Decrement"
+	Add           string = "Add"
+	AddUnique     string = "AddUnique"
+	IncrementSet  string = "IncrementSet"
+	IncrementFrom string = "IncrementFrom"
+)
+
+// ValidateOperation checks that the operation is valid
+func ValidateOperation(p search.PartialUpdateOperation) error {
```
Comment (loicsay) [inline pkg/cmd/objects/update/object.go:28]: Should we create a unit test for that function?

File: pkg/cmd/objects/update/object.go [hunk 6]
```diff
@@ -1,78 +1,78 @@
...
-						Value:     operationValue,
-					}
-				}
-			} else if !isValueString {
-				return fmt.Errorf("Error at object %d: objectID should be a string", objectIndex+1)
-			} else {
-				objectToUpdate["objectID"] = valueString
+	// Each field could be either an `search.PartialUpdateOperation` or a scalar value
+	for k, v := range *o {
+		switch v := v.(type) {
```
Comment (loicsay) [inline pkg/cmd/objects/update/object.go:63]: Elegant way to check if it's a scalar or an operation 👍

File: pkg/cmd/objects/update/object.go (hunk 2 above)
Comment (clemfromspace) [reply pkg/cmd/objects/update/object.go]: 👍

File: pkg/cmd/objects/update/object.go (hunk 5 above)
Comment (clemfromspace) [reply pkg/cmd/objects/update/object.go:28]: Yep 👍

File: pkg/cmd/objects/update/update.go [hunk 7]
```diff
@@ -74,44 +85,110 @@ func NewUpdateCmd(f *cmdutil.Factory, runF func(*UpdateOptions) error) *cobra.Co
...
+			}
+			if !confirmed {
+				return nil
+			}
+		}
+	}
+
+	// Update the objects
+	opts.IO.StartProgressIndicatorWithLabel(fmt.Sprintf("Updating %s objects on %s", cs.Bold(fmt.Sprint(len(objects))), cs.Bold(opts.Index)))
+	res, err := index.PartialUpdateObjects(objects, opt.CreateIfNotExists(opts.CreateIfNotExists))
```
[the remark this answers is not in the document]
Comment (clemfromspace) [reply pkg/cmd/objects/update/update.go:176]: Yep, just tested it now and it's working correctly.

**Review by clemfromspace — APPROVED** [review]

**Pull request merged**